graph-gateway/src/network/subgraph_client/indexers_list_paginated_client/inner_client.rs

```diff
@@ -0,0 +1,267 @@
+use std::sync::{atomic::AtomicU64, Arc};
+
+use alloy_primitives::aliases::BlockNumber;
+use parking_lot::Mutex;
+use serde::de::Deserialize;
+use thegraph_core::types::BlockPointer;
+use thegraph_graphql_http::{graphql::Document, http_client::ResponseError};
+use url::Url;
+
+use super::queries::{
+    meta::send_bootstrap_meta_query,
+    page::{send_subgraph_page_query, BlockHeight, SubgraphPageQueryResponseOpaqueEntry},
+};
+
+/// Error message returned by the indexer typically when a reorg happens.
+const SUBGRAPH_REORG_ERROR: &str = "no block with that hash found";
+
+/// Errors that can occur while sending a paginated query.
+#[derive(Debug, Clone, thiserror::Error)]
+pub enum PaginatedQueryError {
+    /// The bootstrap meta query failed.
+    #[error("bootstrap meta query failed: {0}")]
+    BootstrapMetaQueryFailed(String),
+
+    /// The page response was empty.
+    ///
+    /// A page query response should always contain at least the 'meta' field response. If the
+    /// response is empty, it means that the subgraph is not returning any data.
+    #[error("empty response")]
+    EmptyResponse,
+
+    /// A reorg was detected.
+    ///
+    /// The indexer responded with an error message indicating that a reorg was detected.
+    #[error("reorg detected")]
+    ReorgDetected,
+
+    /// An error occurred while sending one of the requests.
+    #[error("request error: {0}")]
+    RequestError(String),
+
+    /// An error occurred while processing the query.
+    ///
+    /// This error contains the error messages returned by the indexer when an error occurred while
+    /// processing one of the page requests.
+    #[error("response error: {0:?}")]
+    ResponseError(Vec<String>),
+
+    /// Response deserialization error.
+    ///
+    /// An error occurred while deserializing the response.
+    #[error("deserialization error: {0}")]
+    DeserializationError(String),
+}
+
+#[allow(clippy::too_many_arguments)]
+async fn send_paginated_query<T>(
+    client: reqwest::Client,
+    subgraph_url: Url,
+    auth: &str,
+    query: Document,
+    page_size: usize,
+    mut block_height: BlockHeight,
+) -> anyhow::Result<(Vec<T>, Option<BlockPointer>)>
+where
+    T: for<'de> Deserialize<'de> + Send,
+{
+    debug_assert_ne!(page_size, 0, "page size must be greater than 0");
+
+    // Block at which the query is executed.
+    let mut block_pointer: Option<BlockPointer> = None;
+
+    // The last id of the previous batch.
+    let mut last_id: Option<String> = None;
+
+    // Vector to store the results of the paginated query.
+    let results = Mutex::new(Vec::new());
```

Comment:
This mutex seems unnecessary

Reply:
Yes, no longer needed. This is a leftover from the previous implementation where it was necessary to make `Send` the returned `Future`.